ethereum/eth/src/main/java/tech/pegasys/pantheon/ethereum/eth/manager/AbstractFanOutTask.java

```diff
@@ -0,0 +1,161 @@
+
+  private BlockingQueue<I> inboundQueue;
+  private BlockingQueue<O> outboundQueue;
+  private Queue<InFlightRequestData<I, O>> inFlightQueue;
+  private List<O> results;
+
+  private boolean shuttingDown = false;
+  private AtomicReference<Throwable> processingException = new AtomicReference<>(null);
+
+  protected AbstractFanOutTask(
+      final BlockingQueue<I> inboundQueue,
+      final int outboundBacklogSize,
+      final LabelledMetric<OperationTimer> ethTasksTimer) {
+    super(ethTasksTimer);
+    this.inboundQueue = inboundQueue;
+    outboundQueue = new LinkedBlockingQueue<>(outboundBacklogSize);
+    inFlightQueue = new ArrayDeque<>(outboundBacklogSize);
+    results = new ArrayList<>();
+  }
+
+  @Override
+  protected void executeTask() {
+    Optional<I> previousInput = Optional.empty();
+
+    try {
+      // Loop until an exception happened or we have no work and we are told to stop.
+      while ((processingException.get() == null)
+          && !(inboundQueue.isEmpty() && inFlightQueue.isEmpty() && shuttingDown)) {
+        // take all current inbound items and start them working
+        while (!inboundQueue.isEmpty()) {
+          final I currentInput = inboundQueue.poll();
+          final Optional<CompletableFuture<O>> future =
+              startProcessing(currentInput, previousInput);
+          if (future.isPresent()) {
+            inFlightQueue.add(new InFlightRequestData<>(previousInput, currentInput, future.get()));
+          }
+          previousInput = Optional.of(currentInput);
+        }
+
+        // read all in flight processes and post process them
+        while (!inFlightQueue.isEmpty()) {
+          final InFlightRequestData<I, O> inFlightData = inFlightQueue.peek();
+          if (!inFlightData.future.isDone()) {
+            break;
+          }
+          final O futureResult;
+          try {
+            futureResult = inFlightData.future.get();
+          } catch (final InterruptedException e) {
+            // Try Again.
+            continue;
+          } catch (final ExecutionException e) {
+            // Fail completely.
+            result.get().completeExceptionally(e.getCause());
+            return;
+          }
+          // Success. Take it out of the queue.
+          inFlightQueue.poll();
+          final Optional<O> output =
+              finishProcessing(inFlightData.current, inFlightData.previous, futureResult);
+          output.ifPresent(
+              o -> {
+                try {
+                  outboundQueue.put(o);
+                } catch (final InterruptedException e) {
+                  processingException.compareAndSet(null, e);
+                }
+                results.add(o);
+              });
+        }
+
+        // Sleep if inbound queue is empty.
+        if (inboundQueue.isEmpty()) {
+          if (inFlightQueue.isEmpty()) {
+            // If both queues are empty park it for a fixed time
+            LockSupport.parkNanos(WAIT_NS);
```

Comment:
It's a mess of problems to try that:
(a) The only blocking methods are poll and take,  Since all I have is a queue and not a deque there is no way to "put it back" once I get it except hokey "maybe I peeked" fields.
(b) we would want timed polls anyway to check for cancellation of the task.
(c) we don't "own" the inbound queue, so we cannot ensure it gets interrupted or cancelled or can even guarantee the class so we can get the specific lock inside the deeper implementation.
(d) there is no API in the interface that will cause it to send interrupts to all blocked listeners.